MuxExoPlayer/src/r2_12_1/java/com/mux/stats/sdk/muxstats/MuxStatsExoPlayer.java

```diff
@@ -0,0 +1,379 @@
+                }
+                break;
+            case Player.STATE_IDLE:
+            default:
+                // don't care
+                break;
+        }
+    }
+
+    @Override
+    public void onPlayerError(ExoPlaybackException e) {
+        if (e.type == ExoPlaybackException.TYPE_RENDERER) {
+            Exception cause = e.getRendererException();
+            if (cause instanceof MediaCodecRenderer.DecoderInitializationException) {
+                MediaCodecRenderer.DecoderInitializationException die = (MediaCodecRenderer.DecoderInitializationException) cause;
+                if (die.codecInfo != null && die.codecInfo.name == null) {
+                    if (die.getCause() instanceof MediaCodecUtil.DecoderQueryException) {
+                        internalError(new MuxErrorException(e.type, "Unable to query device decoders"));
+                    } else if (die.secureDecoderRequired) {
+                        internalError(new MuxErrorException(e.type, "No secure decoder for " + die.mimeType));
+                    } else {
+                        internalError(new MuxErrorException(e.type, "No decoder for " + die.mimeType));
+                    }
+                } else {
+                    internalError(new MuxErrorException(e.type, "Unable to instantiate decoder for " + die.mimeType));
+                }
+            }
+        } else if (e.type == ExoPlaybackException.TYPE_SOURCE) {
+            Exception error = e.getSourceException();
+            internalError(new MuxErrorException(e.type, error.getClass().getCanonicalName() + " - " + error.getMessage()));
+        } else if (e.type == ExoPlaybackException.TYPE_UNEXPECTED) {
+            Exception error = e.getUnexpectedException();
+            internalError(new MuxErrorException(e.type, error.getClass().getCanonicalName() + " - " + error.getMessage()));
+        } else {
+            internalError(e);
+        }
+    }
+
+    @Override
+    public void onPlayWhenReadyChanged(boolean playWhenReady, int reason) {
+        this.playWhenReady = playWhenReady;
+    }
+
+    @Override
+    public void onPositionDiscontinuity(int reason) {
+        if (reason == Player.DISCONTINUITY_REASON_SEEK) {
+            seeked();
+        }
+    }
+
+    @Override
+    public void onRepeatModeChanged(int repeatMode) { }
+
+    // Note, onSeekProcessed was deprecated in 2.12.0
+
+    @Override
+    public void onShuffleModeEnabledChanged(boolean shuffleModeEnabled) { }
+
+    @Override
+    public void onTimelineChanged(Timeline timeline, int reason) {
+        if (timeline != null && timeline.getWindowCount() > 0) {
+            Timeline.Window window = new Timeline.Window();
+            timeline.getWindow(0, window);
+            sourceDuration = window.getDurationMs();
+        }
+    }
+
+    @Override
+    public void onTracksChanged(TrackGroupArray trackGroups, TrackSelectionArray trackSelections) {
+        bandwidthDispatcher.onTracksChanged(trackGroups);
```

Comment:
Need to add `configurePlaybackHeadUpdateInterval();` so that this will work with the new seeking/rebuffering/playback position tracking.